aten/src/ATen/native/Distance.cpp

```diff
@@ -239,19 +239,67 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
   return result;
 }
 
-Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double eps) {
-  auto common_size = at::infer_size_dimvector(x1.sizes(), x2.sizes());
-  auto commonDtype = at::result_type(x1, x2);
+Tensor cosine_similarity(const Tensor& x1_, const Tensor& x2_, int64_t dim, double eps) {
+  /*
+   * cosine_similarity(x1, x2) = <x1, x2> / (||x1|| * ||x2||)
+   *
+   * The current implementation is an improvement over the previous version.
+   *
+   * Previous implementation:
+   * 1. Compute num = <x1, x2>,
+   * 2. Compute denom = ||x1|| * ||x2||,
+   * 3. Compute denom = max(denom, eps) to avoid division by zero,
+   * 4. Return num / denom.
+   *
+   * Previous implementation has the following issues:
+   * 1. Chance of losing precision in <x1, x2> when ||x1|| and ||x2|| are large.
+   * 2. Chance of losing precision in ||x1|| * ||x2|| when ||x1|| and ||x2|| are large.
+   * 3. Losing precision may cause |cosing_similarity(x1, x2)| > 1.0.
+   *
+   * Current implementation:
+   * 1. Compute x1_normalized = x1 / max(||x1||, eps),
+   *            x2_normalized = x2 / max(||x2||, eps),
+   * 2. Return <x1_normalized, x2_normalized>.
+   *
+   * The current implementation improves over the previous one by:
+   * 1. Making sure that <x1, x2> and ||x1|| * ||x2|| are not computed explicitly,
+   *    hence avoiding floating point overflows.
+   * 2. Both methods might have issues with computing ||x1|| and ||x2||, but for
+   *    the current method this is the only source of the floating point imprecision.
+   * 3. Makes sure |cosing_similarity(x1, x2)| <= 1.0.
+   *
+   */
+  auto commonDtype = at::result_type(x1_, x2_);
   TORCH_CHECK(at::isFloatingType(commonDtype), "expected common dtype to be floating point, yet common dtype is ", commonDtype);
-  Tensor x1_ = x1.to(commonDtype).expand(common_size);
-  Tensor x2_ = x2.to(commonDtype).expand(common_size);
-  // Follow scipy impl to improve numerical precision
-  // Use x / sqrt(x * x) instead of x / (sqrt(x) * sqrt(x))
-  Tensor w12 = at::sum(x1_ * x2_, dim);
-  Tensor w1 = at::sum(x1_ * x1_, dim);
-  Tensor w2 = at::sum(x2_ * x2_, dim);
-  Tensor n12 = (w1 * w2).clamp_min_(eps * eps).sqrt_();
-  return w12.div_(n12);
+
+  auto common_size = at::infer_size_dimvector(x1_.sizes(), x2_.sizes());
+  auto x1 = x1_.to(commonDtype).expand(common_size);
+  auto x2 = x2_.to(commonDtype).expand(common_size);
+
+  auto x1_squared_norm = at::pow(x1, 2).sum(dim, /*keepdim=*/true);
+  auto x2_squared_norm = at::pow(x2, 2).sum(dim, /*keepdim=*/true);
+
+  {
+    at::NoGradGuard guard;
+    x1_squared_norm.clamp_min_(eps * eps);
+    x2_squared_norm.clamp_min_(eps * eps);
+  }
+
+  auto x1_norm = x1_squared_norm.sqrt_();
```

Comment:
I don't remember all the details on the `eps` logic, but why not use `at::norm(x1)` here?

Reply:
We had a discussion about that :)
https://github.com/pytorch/pytorch/pull/31378#issuecomment-572073379

Reply:
Thanks for the link!
But norm is differentiable at 0 today! So not sure what's up...
This code runs just fine:
```python
import torch

def f(x):
    y = x.norm()
    with torch.no_grad():
        y.clamp_min_(0.001)

    return x / y


x = torch.rand(2, requires_grad=True)
out = f(x)
print(out)
out.sum().backward()
print(x.grad)

x = torch.zeros(2, requires_grad=True)
out = f(x)
print(out)
out.sum().backward()
print(x.grad)


```

Reply:
I think another reason not to use norm is accuracy. In the long history of cosine similarity there was a period where it was implemented with norms, and people complained about it being larger than 1. Maybe there's some clever norm formulation that avoids that, I don't know.

Reply:
I'm sure we could also discuss using `rsqrt` and `mul`? :p 

@ngimel in your experience, what is a good principled way to answer the question of which one we want to use?

Reply:
Norm could be used, but it will do essentially the same code minus checks and branching. The backward should be as simple as the backward for norm I guess?

Reply:
Good point, @albanD , that and a place or two where we could do in-place operations.

Reply:
`norm` is much faster as it won't materialize x*x, so if it's differentiable at 0 I suggest we use it. This implementation seems already less efficient than previous and launches many more kernels.

Reply:
All right, how about replacing mul + sum with a single matrix product? I guess there were some efforts to dispatch matmul to matrix-vector when possible.

Reply:
Alas, I cannot use norm efficiently. I cannot clamp norm as gradcheck detects version mismatch, and if I omit clamping between the product and the sqrt, we get nans in my tests. This indicates that norm is still problematic around zero. What shall we do? Go with more precise but less performant implementation or just keep the existing one?

Reply:
I can only offer optimizations if the inputs are 1d/2d with dim=-1 (dim=-2, I guess, we can optimize as well) such that we could use matrix products and not materialize any elementwise-products.

Reply:
Haha, good job, math libraries.